### ISSUE TITLE
"Unsupported converter" for completely unanotated function

### ISSUE THREAD

**Issue by dimaqq** [python/mypy#6172]
A slightly different case with same error message just popped up with mypy==0.650:

```py
import attr


def foo(value):
    return value.split()


@attr.s
class Bar:
    field = attr.ib(default=None, converter=foo)
```

```bash
> mypy test_mypy.py
test_mypy.py:10: error: Unsupported converter, only named functions and types are currently supported
```

**Comment by ilevkivskyi** [comment]
Note that adding at least a single annotation to `foo` fixed this. It looks like this should be easy to fix.

cc @euresti

**Comment by Infernion** [comment]
How is possible to skip this warning? It appears even in typed functions and it annoys me.

**Comment by gvanrossum** [comment]
Someone (you?) will have to update the plugin.

**Comment by officialcryptomaster** [comment]
@ilevkivskyi can you demonstrate how I Can add an annotation to suppress this error?

**Comment by officialcryptomaster** [comment]
Ah @ilevkivskyi I see you mean this:

```
import attr

def foo(value) -> str:
    return value.split()

@attr.s
class Bar:
    field = attr.ib(default=None, converter=foo)
```

**Comment by ilevkivskyi** [comment]
@officialcryptomaster Yes.

**Comment by officialcryptomaster** [comment]
Seems like similar issue exists with flake8 which does not go away with type annotations... Any idea how I can work around that one other than suppress it?

**Comment by JelleZijlstra** [comment]
@officialcryptomaster this sounds like you should report the issue to flake8, unless I'm misunderstanding what you say.

**Comment by webknjaz** [comment]
This is also happening with annotated functions that use [`functools.singledispatch()`](https://docs.python.org/3/library/functools.html#functools.singledispatch).

**Comment by madig** [comment]
Maybe the warning should read "Unsupported converter, only types and named functions with an annotated return type are currently supported" 🙂

**Comment by JarnoRFB** [comment]
This somehow seems to be broken for `attrs 20.3`. For 

```python
import attr

def foo(value) -> str:
    return value.split()

@attr.s
class Bar:
    field = attr.ib(default=None, converter=foo)
```

`mypy` gives me ` error: Need type annotation for 'field'`, while it works with `attrs` `19.3`. Not sure if this issue needs to be addressed in `mypy` or `attrs`.

**Comment by euresti** [comment]
I'm pretty sure this last one is because `attrs` changed the stubs.

**Comment by JarnoRFB** [comment]
@euresti It still works if one annotates the attribute itself, just was unaware of this possibility, because pycharm issues a misleading warning. 

```python
def foo(value: str) -> List[str]:
    return value.split()


@attr.s
class Bar:
    field: List[str] = attr.ib(default=None, converter=foo)


bar = Bar("5 5")
reveal_type(bar.field) # Revealed type is 'builtins.list[builtins.str]'
```

**Comment by DiddiLeija** [comment]
Hello! Is this still happening? If I can make a contribution, just tell me where to start.

**Comment by cgebbe** [comment]
@DiddiLeija: Error message originates from here: https://github.com/python/mypy/blob/master/mypy/plugins/attrs.py#L623

**Comment by DiddiLeija** [comment]
In simple words, we have to provide a friendlier message, right? Or should I move something in the behavior?

**Comment by JelleZijlstra** [comment]
I feel like this should just work and infer `Any` as the input type.

**Comment by cgebbe** [comment]
@JelleZijlstra : I second this. In my understanding this makes type checking optional.

If functions have annotations, types should be checked. If functions have no annotations, (effectively) no type checking is performed on them.

**Comment by xqgex** [comment]
I'm not sure if this is related here or not, I got the same error when using attrs converter with classmethod.
`mypy` version `1.1.1` and `attrs` version: `23.1.0`

```Python
from typing import Any

from attrs import define, field


class Foo(str):
    @classmethod
    def from_bar(cls, value: Any) -> 'Foo':
        return cls(str(value))


@define
class Bar:
    field: Foo = field(converter=Foo.from_bar)


bar = Bar("5 5")
reveal_type(bar.field)
```

Output:

```
test.py:14: error: Unsupported converter, only named functions, types and lambdas are currently supported  [misc]
test.py:18: note: Revealed type is "test.Foo"
Found 1 error in 1 file (checked 1 source file)
```
